substrate/primitives/core/src/address_uri.rs

```diff
@@ -0,0 +1,302 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Little util for parsing an address URI. Replaces regular expressions.
+
+/// A container for results of parsing the address uri string.
+///
+/// Intended to be equivalent of:
+/// `Regex::new(r"^(?P<phrase>[a-zA-Z0-9 ]+)?(?P<path>(//?[^/]+)*)(///(?P<password>.*))?$")`
+/// which also handles soft and hard derivation paths:
+/// `Regex::new(r"/(/?[^/]+)")`
+///
+/// Example:
+/// ```
+/// 	use sp_core::crypto::AddressUri;
+/// 	let manual_result = AddressUri::parse("hello world/s//h///pass");
+/// 	assert_eq!(
+/// 		manual_result.unwrap(),
+/// 		AddressUri { phrase: Some("hello world"), paths: vec!["s", "/h"], pass: Some("pass") }
+/// 	);
+/// ```
+#[derive(Debug, PartialEq)]
+pub struct AddressUri<'a> {
+	/// Phrase, hexadecimal string, or ss58-compatible string.
+	pub phrase: Option<&'a str>,
+	/// Key derivation paths, ordered as in input string,
+	pub paths: Vec<&'a str>,
+	/// Password.
+	pub pass: Option<&'a str>,
+}
+
+/// Errors that are possible during parsing the address URI.
+#[allow(missing_docs)]
+#[derive(Debug, thiserror::Error, PartialEq, Eq, Clone, Copy)]
+pub enum Error {
+	#[error("Invalid character in phrase")]
+	InvalidCharacterInPhrase,
+	#[error("Invalid character in password")]
+	InvalidCharacterInPass,
+	#[error("Invalid character in hard path")]
+	InvalidCharacterInHardPath,
+	#[error("Invalid character in soft path")]
+	InvalidCharacterInSoftPath,
+}
+
+impl<'a> AddressUri<'a> {
+	fn extract_prefix(input: &mut &'a str, is_allowed: &dyn Fn(char) -> bool) -> Option<&'a str> {
+		let output = input.trim_start_matches(is_allowed);
+		let prefix_len = input.len() - output.len();
+		let prefix = if prefix_len > 0 { Some(&input[..prefix_len]) } else { None };
+		*input = output;
+		prefix
+	}
+
+	/// Parses the given string.
+	pub fn parse(mut input: &'a str) -> Result<Self, Error> {
+		let phrase = Self::extract_prefix(&mut input, &|ch: char| {
+			ch.is_ascii_digit() || ch.is_ascii_alphabetic() || ch == ' '
+		});
+
+		let mut pass = None;
+		let mut paths = Vec::new();
+		while !input.is_empty() {
+			input = if let Some(mut maybe_pass) = input.strip_prefix("///") {
+				pass = match Self::extract_prefix(&mut maybe_pass, &|ch: char| ch != '\n') {
+					Some(pass) => Some(pass),
+					None => Some(""),
+				};
+				maybe_pass
```

Comment:
You could add another helper to simplify this and other branches a little bit, e.g. maybe something like this:

```rust
fn strip_prefix(input: &mut &str, prefix: &str) -> bool {
    if let Some(new_input) = input.strip_prefix(prefix) {
        *input = new_input;
        true
    } else {
        false
    }
}
```

And then this snippet could be:

```suggestion
			if strip_prefix(&mut input, "///") {
				pass = Some(Self::extract_prefix(&mut input, &|ch: char| ch != '\n').unwrap_or(""));
```

Reply:
I like the idea, code looks simpler. The first `if branch` does not fit here however, since we need original input. I'll try to figure out something.

Reply:
done: 77ad7f1428